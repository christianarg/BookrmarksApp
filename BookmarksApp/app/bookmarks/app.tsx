import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { TagsRoot } from './bookmarks';

class App extends React.Component{
    render() {
        return (<TagsRoot />);
    }
}

ReactDOM.render(<App />, document.getElementById('root'));