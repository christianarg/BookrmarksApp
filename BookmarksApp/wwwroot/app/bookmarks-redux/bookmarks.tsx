import * as React from 'react';
import { BookmarkModel, EditBookmark } from './bookmark-model';
import { AddOrEditBookmark } from "./add-or-edit-bookmark-props";
type BookmarkProps = {
    bookmarks: BookmarkModel[];
    onEdit: (bookrmark: EditBookmark) => void;
};
export function Bookmarks(props: BookmarkProps) {
    const bookmarks = props.bookmarks.map(bookmark => !bookmark.hidden &&
        <li key={bookmark.name} className="bookmark">
            <a href={bookmark.url} target="_blank">{bookmark.name}</a>&nbsp;&nbsp;<AddOrEditBookmark bookmarkToEdit={bookmark} onAddOrEdit={edited => props.onEdit(edited)} />
        </li>);
    return (<ul style={{ listStyleType: 'square' }}>{bookmarks}</ul>);
}
