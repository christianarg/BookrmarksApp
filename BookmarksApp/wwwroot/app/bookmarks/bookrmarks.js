"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
function Bookmarks(props) {
    var bookmarks = props.bookmarks.map(function (bookmark) {
        return React.createElement("li", { key: bookmark.name },
            React.createElement("a", { href: bookmark.url, target: "_blank" }, bookmark.name));
    });
    return (React.createElement("ul", null, bookmarks));
}
function Tag(props) {
    var tag = props.tag;
    var tagItem = React.createElement("li", null,
        tag.name,
        React.createElement(Bookmarks, { bookmarks: tag.bookmarks }));
    return (React.createElement(React.Fragment, null,
        tagItem,
        tag.subTags && React.createElement(Tags, { tags: tag.subTags }),
        " "));
}
function Tags(props) {
    var tagItems = props.tags.map(function (tag) { return React.createElement(Tag, { key: tag.name, tag: tag }); });
    return (React.createElement("ul", null, tagItems));
}
function TagSearch(props) {
    return (React.createElement("input", { type: "text", value: props.searachText, onChange: function (evt) { return props.onSearchChange(evt.target.value); } }));
}
var TagsRoot = /** @class */ (function (_super) {
    __extends(TagsRoot, _super);
    function TagsRoot() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = { tags: null, searachText: '' };
        _this.handleSearch = function (searchValue) {
            _this.setState({ searachText: searchValue });
        };
        return _this;
    }
    TagsRoot.prototype.componentDidMount = function () {
        this.setState({ tags: exports.sampleBookrmarks.slice() });
    };
    TagsRoot.prototype.render = function () {
        if (this.state.tags) {
            return (React.createElement("div", null,
                React.createElement(TagSearch, { searachText: this.state.searachText, onSearchChange: this.handleSearch }),
                React.createElement(Tags, { tags: this.state.tags })));
        }
        return null;
    };
    return TagsRoot;
}(React.Component));
exports.TagsRoot = TagsRoot;
exports.sampleBookrmarks = [
    {
        name: '.Net',
        bookmarks: [
            { name: 'download', url: 'https://dotnet.microsoft.com/download' }
        ],
        subTags: [
            {
                name: 'Asp.net',
                bookmarks: [
                    { name: 'asp.net', url: 'https://dotnet.microsoft.com/apps/aspnet' },
                    { name: 'asp.net core', url: 'https://docs.microsoft.com/es-es/aspnet/core/?view=aspnetcore-2.2' },
                ],
            }
        ]
    },
    {
        name: 'react',
        bookmarks: [
            { name: 'react docs', url: 'https://reactjs.org/docs/getting-started.html' }
        ]
    }
];
//# sourceMappingURL=bookrmarks.js.map