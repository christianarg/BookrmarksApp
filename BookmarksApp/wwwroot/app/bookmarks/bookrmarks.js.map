{"version":3,"file":"bookrmarks.js","sourceRoot":"","sources":["bookrmarks.tsx"],"names":[],"mappings":";;;;;;;;;;;;;;;AAAA,6BAA+B;AAkB/B,SAAS,SAAS,CAAC,KAAmB;IAClC,IAAM,SAAS,GAAG,KAAK,CAAC,SAAS,CAAC,GAAG,CAAC,UAAA,QAAQ;QAC1C,OAAA,4BAAI,GAAG,EAAE,QAAQ,CAAC,IAAI;YAClB,2BAAG,IAAI,EAAE,QAAQ,CAAC,GAAG,EAAE,MAAM,EAAC,QAAQ,IAAG,QAAQ,CAAC,IAAI,CAAK,CAC1D;IAFL,CAEK,CAAC,CAAA;IACV,OAAO,CAAC,gCAAK,SAAS,CAAM,CAAC,CAAC;AAClC,CAAC;AASD,SAAS,GAAG,CAAC,KAAe;IACxB,IAAM,GAAG,GAAG,KAAK,CAAC,GAAG,CAAC;IACtB,IAAM,OAAO,GAAG;QAAK,GAAG,CAAC,IAAI;QAAC,oBAAC,SAAS,IAAC,SAAS,EAAE,GAAG,CAAC,SAAS,GAAI,CAAK,CAAC;IAE3E,OAAO,CAAC;QAAG,OAAO;QAAE,GAAG,CAAC,OAAO,IAAI,oBAAC,IAAI,IAAC,IAAI,EAAE,GAAG,CAAC,OAAO,GAAI;YAAK,CAAC,CAAC;AACzE,CAAC;AAOD,SAAS,IAAI,CAAC,KAAgB;IAC1B,IAAM,QAAQ,GAAG,KAAK,CAAC,IAAI,CAAC,GAAG,CAAC,UAAA,GAAG,IAAI,OAAA,oBAAC,GAAG,IAAC,GAAG,EAAE,GAAG,CAAC,IAAI,EAAE,GAAG,EAAE,GAAG,GAAI,EAAhC,CAAgC,CAAC,CAAC;IACzE,OAAO,CACH,gCAAK,QAAQ,CAAM,CACtB,CAAC;AACN,CAAC;AAQD,SAAS,SAAS,CAAC,KAAqB;IACpC,OAAO,CAAC,+BAAO,IAAI,EAAC,MAAM,EAAC,KAAK,EAAE,KAAK,CAAC,WAAW,EAAE,QAAQ,EAAE,UAAC,GAAG,IAAK,OAAA,KAAK,CAAC,cAAc,CAAC,GAAG,CAAC,MAAM,CAAC,KAAK,CAAC,EAAtC,CAAsC,GAAI,CAAC,CAAA;AACvH,CAAC;AAOD;IAA8B,4BAAkC;IAAhE;QAAA,qEAsBC;QApBG,WAAK,GAAkB,EAAE,IAAI,EAAE,IAAI,EAAE,WAAW,EAAE,EAAE,EAAE,CAAA;QAMtD,kBAAY,GAAG,UAAC,WAAmB;YAC/B,KAAI,CAAC,QAAQ,CAAC,EAAE,WAAW,EAAE,WAAW,EAAE,CAAC,CAAC;QAChD,CAAC,CAAA;;IAYL,CAAC;IAlBG,oCAAiB,GAAjB;QACI,IAAI,CAAC,QAAQ,CAAC,EAAE,IAAI,EAAE,wBAAgB,CAAC,KAAK,EAAE,EAAE,CAAC,CAAC;IACtD,CAAC;IAMD,yBAAM,GAAN;QACI,IAAI,IAAI,CAAC,KAAK,CAAC,IAAI,EAAE;YACjB,OAAO,CACH;gBACI,oBAAC,SAAS,IAAC,WAAW,EAAE,IAAI,CAAC,KAAK,CAAC,WAAW,EAAE,cAAc,EAAE,IAAI,CAAC,YAAY,GAAI;gBACrF,oBAAC,IAAI,IAAC,IAAI,EAAE,IAAI,CAAC,KAAK,CAAC,IAAI,GAAI,CAC7B,CAAC,CAAC;SACf;QACD,OAAO,IAAI,CAAC;IAChB,CAAC;IACL,eAAC;AAAD,CAAC,AAtBD,CAA8B,KAAK,CAAC,SAAS,GAsB5C;AAtBY,4BAAQ;AAwBR,QAAA,gBAAgB,GAAe;IACxC;QACI,IAAI,EAAE,MAAM;QACZ,SAAS,EAAE;YACP,EAAE,IAAI,EAAE,UAAU,EAAE,GAAG,EAAE,uCAAuC,EAAE;SACrE;QACD,OAAO,EAAE;YACL;gBACI,IAAI,EAAE,SAAS;gBACf,SAAS,EAAE;oBACP,EAAE,IAAI,EAAE,SAAS,EAAE,GAAG,EAAE,0CAA0C,EAAE;oBACpE,EAAE,IAAI,EAAE,cAAc,EAAE,GAAG,EAAE,mEAAmE,EAAE;iBACrG;aACJ;SACJ;KACJ;IACD;QACI,IAAI,EAAE,OAAO;QACb,SAAS,EAAE;YACP,EAAE,IAAI,EAAE,YAAY,EAAE,GAAG,EAAE,+CAA+C,EAAE;SAC/E;KACJ;CACJ,CAAA"}