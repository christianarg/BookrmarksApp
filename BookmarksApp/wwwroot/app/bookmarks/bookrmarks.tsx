import * as React from 'react';
import { url } from 'inspector';

export type TagModel = {
    name: string;
    subTags?: TagModel[];
    bookmarks: BookmarkModel[];
    hidden?: boolean;
}

export type BookmarkModel = {
    name: string;
    url: string;
    hidden?: boolean;
}


type BookmarProps = {
    bookmarks: BookmarkModel[];
}

function Bookmarks(props: BookmarProps) {
    const bookmarks = props.bookmarks.map(bookmark => !bookmark.hidden &&
        <li key={bookmark.name} className="bookmark">
            <a href={bookmark.url} target="_blank" >{bookmark.name}</a>
        </li>)
    return (<ul>{bookmarks}</ul>);
}

type AddBookmarkProps = { onAdd: (newBookmark: BookmarkModel) => void };
type AddBookmarkState = { isFormVisible: boolean, name: string, url: string };

class AddBookmark extends React.Component<AddBookmarkProps, AddBookmarkState>{
    state: AddBookmarkState = { isFormVisible: false, name: null, url: null };

    hasValue() {
        const { name, url } = this.state;
        return (name && url);
    }

    handleSubmit = (evt: React.FormEvent) => {
        evt.preventDefault();
        const { name, url } = this.state;
        if (name && url) {
            this.props.onAdd({ name: name, url: url });
        }
        this.setState({ isFormVisible: false });
    }

    handleNameChange = (evt: React.ChangeEvent<HTMLInputElement>) => {
        this.setState({ name: evt.target.value });
    }

    handleUrlChange = (evt: React.ChangeEvent<HTMLInputElement>) => {
        this.setState({ url: evt.target.value });
    }

    toggleShow = () => {
        this.setState({ isFormVisible: !this.state.isFormVisible });
    }

    buttonStyle(): React.CSSProperties {
        if (!this.hasValue()) {
            return { cursor: 'not-allowed' };
        }
        return null;
    }

    render() {
        const buttonStyle = { display: 'inline-block', textDecoration: 'underline', cursor: 'pointer' };
        if (this.state.isFormVisible) {
            return (
                <div>
                    <div style={buttonStyle} onClick={this.toggleShow}>Close</div>
                    <form onSubmit={this.handleSubmit}>
                        <div>Name: <input type="text" onChange={this.handleNameChange} placeholder="Bookmark name..." /></div>
                        <div>Url:  <input type="text" onChange={this.handleUrlChange} placeholder="url..." /></div>
                        <button value="Add" style={this.buttonStyle()}>Add</button>
                    </form>
                </div>);
        }
        return <div style={buttonStyle} onClick={this.toggleShow}>(Add Bookmark)</div>
    }
}


type AddTagProps = { onAdd: (newTag: TagModel) => void };
type AddTagState = { isFormVisible: boolean, name: string };

class AddTag extends React.Component<AddTagProps, AddTagState>{
    state: AddTagState = { isFormVisible: false, name: null };

    hasValue() {
        return this.state.name != null;
    }

    handleSubmit = (evt: React.FormEvent) => {
        evt.preventDefault();
        const { name } = this.state;
        if (name) {
            this.props.onAdd({ name: name, bookmarks: [] });
        }
        this.setState({ isFormVisible: false });
    }

    handleNameChange = (evt: React.ChangeEvent<HTMLInputElement>) => {
        this.setState({ name: evt.target.value });
    }

    toggleShow = () => {
        this.setState({ isFormVisible: !this.state.isFormVisible });
    }

    buttonStyle(): React.CSSProperties {
        if (!this.hasValue()) {
            return { cursor: 'not-allowed' };
        }
        return null;
    }

    render() {
        const buttonStyle = { display: 'inline-block', textDecoration: 'underline', cursor: 'pointer' };
        if (this.state.isFormVisible) {
            return (
                <div>
                    <div style={buttonStyle} onClick={this.toggleShow}>Close</div>
                    <form onSubmit={this.handleSubmit}>
                        <div>Name: <input type="text" onChange={this.handleNameChange} placeholder="Tag name..." /></div>
                        <button value="Add" style={this.buttonStyle()}>Add</button>
                    </form>
                </div>);
        }
        return <div style={buttonStyle} onClick={this.toggleShow}>(Add Tag)</div>
    }
}

type TagProps = {
    tag: TagModel;
    onAddBookmark: (tag: TagModel, bookmark: BookmarkModel) => void;
    onAddTag: (newTag: TagModel, parentTag: TagModel) => void;
}

function Tag(props: TagProps) {
    const tag = props.tag;
    if (tag.hidden) {
        return null;
    }

    return (
        <li className="tag">
            <span>{tag.name}</span> <AddBookmark onAdd={(newBookmark) => props.onAddBookmark({ ...tag }, newBookmark)} />
            <Bookmarks bookmarks={tag.bookmarks} />
            {tag.subTags && <Tags tags={tag.subTags} onAddBookmark={props.onAddBookmark} onAddTag={props.onAddTag} />}
            <AddTag onAdd={(newTag) => props.onAddTag(newTag, tag)} />
        </li>);
}


type TagsProps = {
    tags: TagModel[];
    onAddBookmark: (tag: TagModel, bookmark: BookmarkModel) => void;
    onAddTag: (newTag: TagModel, parentTag: TagModel) => void;
}

function Tags(props: TagsProps) {
    const tagItems = props.tags.map(tag => <Tag key={tag.name} tag={tag} onAddBookmark={props.onAddBookmark} onAddTag={props.onAddTag} />);
    return (
        <ul>{tagItems}</ul>
    );
}


type TagSearchProps = {
    searachText: string;
    onSearchChange: (value: string) => void;
}

function TagSearch(props: TagSearchProps) {
    const { searachText, onSearchChange } = props;
    return (<div><input type="text" placeholder="Search tags..." value={searachText} onChange={(evt) => onSearchChange(evt.target.value)} /></div>)
}


function hasText(text: string, searchText: string) {
    return text.toLowerCase().includes(searchText.toLowerCase());
}

export function filterTags(tags: TagModel[], searchText: string) {
    tags.forEach(tag => {
        tag.bookmarks.forEach(bookmark => bookmark.hidden = !hasText(bookmark.name, searchText));
        const anyBookmarkVisible = tag.bookmarks.some(b => !b.hidden);
        const tagNameHasText = hasText(tag.name, searchText);
        const anySubTagsVisible = tag.subTags && tag.subTags.some(t => !t.hidden);
        tag.hidden = !tagNameHasText && !anyBookmarkVisible && !anySubTagsVisible;
        if (tag.subTags) {
            filterTags(tag.subTags, searchText);
        }
    });
}

export function replaceTag(tags: TagModel[], newTag: TagModel): TagModel[] {
    const newTags = tags.map(tag => tag.name == newTag.name ? newTag : tag);

    newTags.forEach(tag => {
        if (tag.subTags) {
            tag.subTags = replaceTag(tag.subTags, newTag);
        }
    });
    return newTags;
}

type TagsRootState = {
    tags: TagModel[];
    searachText: string;
}

export class TagsRoot extends React.Component<{}, TagsRootState> {

    state: TagsRootState = { tags: null, searachText: '' }

    componentDidMount() {
        this.setState({ tags: sampleBookrmarks.slice() });
    }

    handleSearch = (searchValue: string) => {
        const tags = this.state.tags.slice();
        filterTags(tags, searchValue);
        this.setState({ tags: tags, searachText: searchValue });
    }

    handleAddBookmark = (tag: TagModel, bookmarkModel: BookmarkModel) => {
        let tags = this.state.tags.slice();
        tag.bookmarks.push(bookmarkModel);

        tags = replaceTag(tags, tag);
        this.setState({ tags: tags });
    }

    handleAddTag = (newTag: TagModel, parentTag: TagModel) => {
        let tags = this.state.tags.slice();
        if (parentTag) {
            parentTag.subTags.push(newTag);
            replaceTag(tags, { ...parentTag });
        }
        else {
            // sin parent
            tags.push(newTag);
        }
        this.setState({ tags: tags });
    }

    render() {
        if (this.state.tags) {
            return (
                <div>
                    <TagSearch searachText={this.state.searachText} onSearchChange={this.handleSearch} />
                    <Tags tags={this.state.tags} onAddBookmark={this.handleAddBookmark} onAddTag={(newTag) => this.handleAddTag(newTag, null)} />
                </div>);
        }
        return null;
    }
}

export const sampleBookrmarks: TagModel[] = [
    {
        name: '.Net',
        bookmarks: [
            { name: 'download', url: 'https://dotnet.microsoft.com/download' }
        ],
        subTags: [
            {
                name: 'Asp.net',
                bookmarks: [
                    { name: 'asp.net', url: 'https://dotnet.microsoft.com/apps/aspnet' },
                    { name: 'asp.net core', url: 'https://docs.microsoft.com/es-es/aspnet/core/?view=aspnetcore-2.2' },
                ],
            }
        ]
    },
    {
        name: 'react',
        bookmarks: [
            { name: 'react docs', url: 'https://reactjs.org/docs/getting-started.html' }
        ]
    }
]