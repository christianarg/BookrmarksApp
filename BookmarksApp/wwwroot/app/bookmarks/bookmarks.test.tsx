import * as Bookrmarks from './bookmarks';

describe('filterTests', () => {
    test('filter', () => {
        // ARRANGE
        const tag1: Bookrmarks.TagModel = {
            name: '.Net',
            bookmarks: [
                { name: 'download', url: 'https://dotnet.microsoft.com/download' }
            ]
        };
        const tag2: Bookrmarks.TagModel = {
            name: 'react',
            bookmarks: [
                { name: 'react docs', url: 'https://reactjs.org/docs/getting-started.html' }
            ]
        };
        const tags = [tag1, tag2];

        // ACT
        Bookrmarks.filterTags([tag1, tag2], 'down');

        // ARRANGE
        expect(tag1.hidden).toBe(false);
        expect(tag1.bookmarks[0].hidden).toBe(false);

        expect(tag2.hidden).toBe(true);
        expect(tag2.bookmarks[0].hidden).toBe(true);

    });
})

describe('replaceTagTests', () => {
    test('replaceTag', () => {
        // ARRANGE
        const newTag: Bookrmarks.TagModel = {
            name: '.Net',
            bookmarks: [{ name: 'download', url: 'https://dotnet.microsoft.com/download' }]
        }
        const tag1: Bookrmarks.TagModel = {
            name: '.Net',
            bookmarks: []
        };
        const tag2: Bookrmarks.TagModel = {
            name: 'react',
            bookmarks: []
        };
        const tags = [tag1, tag2];
        // ACT
        const replacedTags = Bookrmarks.replaceTag(tags, newTag);

        // ASSERT
        var newTagFromReplaced = replacedTags.find(tag => tag.name == newTag.name);
        expect(newTagFromReplaced.bookmarks[0].name).toBe(newTag.bookmarks[0].name);
    });

    test('replaceSubTag', () => {
        // ARRANGE
        const newTag: Bookrmarks.TagModel = {
            name: 'Asp.Net',
            bookmarks: [ { name: 'asp.net', url: 'https://dotnet.microsoft.com/apps/aspnet' }]
        }

        const subTag: Bookrmarks.TagModel = {
            name: 'Asp.Net',
            bookmarks: []
        };

        const tag1: Bookrmarks.TagModel = {
            name: '.Net',
            bookmarks: [],
            subTags: [subTag]
        };
        
        const tags = [tag1];
        // ACT
        const replacedTags = Bookrmarks.replaceTag(tags, newTag);

        // ASSERT
        var newTagFromReplaced = replacedTags[0].subTags.find(tag => tag.name == newTag.name);
        expect(newTagFromReplaced.bookmarks[0].name).toBe(newTag.bookmarks[0].name);
    });
})