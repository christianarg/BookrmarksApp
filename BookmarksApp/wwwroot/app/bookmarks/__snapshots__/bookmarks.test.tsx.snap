// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsRoot tests Render root tag test 1`] = `
<div>
  <TagSearch
    onSearchChange={[Function]}
    searachText=""
  />
  <Tags
    onAddBookmark={[Function]}
    onAddTag={[Function]}
    onEditBookmark={[Function]}
    parentTag={null}
    tags={
      Array [
        Object {
          "bookmarks": Array [
            Object {
              "name": "download",
              "url": "https://dotnet.microsoft.com/download",
            },
          ],
          "name": ".Net",
          "subTags": Array [
            Object {
              "bookmarks": Array [
                Object {
                  "name": "asp.net",
                  "url": "https://dotnet.microsoft.com/apps/aspnet",
                },
                Object {
                  "name": "asp.net core",
                  "url": "https://docs.microsoft.com/es-es/aspnet/core/?view=aspnetcore-2.2",
                },
              ],
              "name": "Asp.net",
            },
          ],
        },
        Object {
          "bookmarks": Array [
            Object {
              "name": "react docs",
              "url": "https://reactjs.org/docs/getting-started.html",
            },
          ],
          "name": "React",
        },
      ]
    }
  />
  <AddOrEditTag
    isRoot={true}
    onAddOrEdit={[Function]}
  />
</div>
`;
